### ISSUE TITLE
Fix false positives for "always" option in declaration-empty-line-before

### ISSUE THREAD

**Issue by larrifax** [stylelint/stylelint#4635]
> Clearly describe the bug

There are false positives for `declaration-empty-line-before` when adding simple styles using `styled-components` with stylelints built-in support for css-in-js. See example & demo below for a reproduction.

Might be related to #4003.

> Which rule, if any, is the bug related to?

`declaration-empty-line-before`

> What code is needed to reproduce the bug?

```css
const StyledDiv = styled.div`
  width: 10rem;
`;
```

> What stylelint configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "declaration-empty-line-before": [
      "always",
      {
        "except": [
          "first-nested"
        ]
      }
    ]
  }
}
```

> Which version of stylelint are you using?

`13.1.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

CLI with `stylelint src/**/*.{ts,tsx}`. The same issue can be observed on the [demo page](https://stylelint.io/demo#N4Igxg9gJgpiBc4IDsDOAXABAZXQTwBsYoARASwDdMBeTDQ4gOikoAMAdZTTAdzKnQALeJgCMABgBOMALYBuTqwXIQAGhCo8ydAEMAHgnCpUAWjLITAK1RqkyAGZkA5oeCdu7EJICuRVJ5E3Lm5MT1gwAh1JHXQyFBNZAAd8EwJzGBMAIxh7CGkAzABtdxCPEB0CHh08fzUS0qDS0s8YPTAYZILi4KaQz0dJDBNkGAxiT3qmgF1J7gBfSZngheQ5kDmgA).

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

Yes, it's related to css-in-js (styled-components).

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

The following errors were flagged:

```shell
Expected empty line before declaration (declaration-empty-line-before)
```

**Comment by larrifax** [comment]
I just noticed #4574. Maybe this bug should be added to the list?

**Comment by jeddy3** [comment]
@larrifax Thanks for the clear report, using the template and for creating a reproducible demo.

The issue is either with:

- our [`isFirstNested` util](https://github.com/stylelint/stylelint/blob/master/lib/utils/isFirstNested.js)
- the parser that [stylelint uses to support this syntax](https://stylelint.io/about/syntaxes)

If it's the former, please consider [contributing](https://stylelint.io/contributing) a fix to stylelint itself. There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide. If it's the latter, please consider [contributing to the parser](https://stylelint.io/developer-guide/syntaxes#working-on-syntaxes).

> I just noticed #4574. Maybe this bug should be added to the list?

Done.

**Comment by larrifax** [comment]
I've done a bit of research, and it seems like the problem is that `isFirstNested` returns `false` for the above mentioned code. The culprit is the following line: https://github.com/stylelint/stylelint/blob/343e40dcb625e4d1001db441e4dcf7ac105784cf/lib/utils/isFirstNested.js#L12-L13

In this specific case, `parentNode.type === 'root'`, which results in `false` being returned. If I remove the root check, the tests I've written passes.
From debugging, I don't see any obvious solution for detecting that this code originates from css-in-js / styled-components, which would allow me to handle this case specially.

From the discussion you had on #4003, I understand that you:
* don't want to skip the root check entirely, as that would mean the code would pick up on "declarations" outside of rules.
* don't want to change `postcss-jsx` to wrap styles in e.g. `div` on parse, as that would lead to false positives on other rules.

The suggested fix at that time was to introduce some kind of `context.isStyled`, but it seems that such a solution was never implemented.

@jeddy3 Do you have comments as to how to proceed further? It seems like a decision needs to be made by someone with a more thorough understanding of the codebase and the vision of stylelint than me on this one.

**Comment by jeddy3** [comment]
@larrifax Many thanks for digging deeper.

I suggest we change `postcss-jsx` (now [`postcss-css-in-js`](https://github.com/stylelint/postcss-css-in-js)) to wrap styles in a rule-set with an unqualified class selector. I think that's a more accurate representation of what the CSS-in-JS source would look like as vanilla CSS.

> and the vision of stylelint than me on this one.

The built-in rules support standard CSS because of the speed of change in the world of CSS-like language extensions and, especially in, CSS-in-JS styling frameworks.

From [the docs](https://stylelint.io/about/syntaxes):

> We lean on PostCSS syntaxes to help us with this task [of supporting CSS-like language extensions and CSS-in-JS styling frameworks]. We use them to transform these languages into something that resembles CSS, which is the language that:
>  - underpins all the other styling languages
>  - is best understood by rules built into stylelint

I think the PostCSS syntaxes should transform the source into an AST that will closely resemble the final CSS. I think that approach would make the built-in rules useful to the most number of people.

---

We're trying to rally community support for maintaining `postcss-css-in-js`, so please don't hesitate to get involved in the parser if you have time.

**Comment by jeddy3** [comment]
Closing [as stylistic rules are frozen](https://github.com/stylelint/stylelint/issues/5674). The community is welcome to migrate the rule [to a plugin](https://stylelint.io/developer-guide/plugins) and fix this bug.
